proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -13958,27 +13883,60 @@ lemma maybe_donate_sc_active_sc_valid_refills[wp]:
                  hoare_vcg_if_lift2)
 
 lemma postpone_ready_or_release[wp]:
-  "\<lbrace>valid_ready_qs and ready_or_release\<rbrace>
+  "\<lbrace>ready_qs_etcb_eq and ready_or_release\<rbrace>
    postpone sc_ptr
    \<lbrace>\<lambda>_. ready_or_release\<rbrace>"
   unfolding postpone_def get_sc_obj_ref_def
   by (wpsimp wp: tcb_dequeue_not_queued)+
 
 lemma sched_context_resume_ready_or_release[wp]:
-  "\<lbrace>valid_ready_qs and ready_or_release\<rbrace>
+  "\<lbrace>ready_qs_etcb_eq and ready_or_release\<rbrace>
    sched_context_resume sc_ptr
    \<lbrace>\<lambda>_. ready_or_release\<rbrace>"
   unfolding sched_context_resume_def
-  by (wpsimp wp: hoare_drop_imps hoare_vcg_all_lift hoare_vcg_if_lift simp: is_schedulable_def)
+  by (wpsimp wp: thread_get_wp is_schedulable_wp)
+     (clarsimp simp: obj_at_def is_schedulable_bool_def2 pred_tcb_at_def is_tcb)
+
+lemma tcb_sched_enqueue_ready_queues_etcb_eq[wp]:
+  "tcb_sched_action tcb_sched_enqueue tp \<lbrace>ready_qs_etcb_eq\<rbrace>"
+  unfolding tcb_sched_action_def tcb_sched_enqueue_def get_tcb_queue_def bind_assoc
+  apply (rule hoare_seq_ext[OF _ thread_get_sp])
```

Comment:
Are these `hoare_seq_ext`s necessary? (And in the next couple proofs)

Reply:
yes, otherwise we’ll need to assume `tcb_at` I think.

> So did you end up using this new abbreviation in the definition of `valid_ready_qs`? If not, that might be nice, if possible

It might be possible eventually when we tweak `valid_ready_qs`, but currently it is a multilayered definition/abbreviation and `\<forall>tp d p. tp \<in> set (ready_queues s d p)` part is in a different layer from the rest of the statement. So it is not quite possible to use this as is. The abbreviation `etcb_eq` does appear in `valid_ready_qs` at some level.

Reply:
Hmm, it's hard to tell without looking at the proofs, but that surprises me. In theory there shouldn't be any difference in power between sp and wp style proofs, and if this is true then I think that that would mean that the wp rules could be improved.

Reply:
You are quite right there, actually. It was because of `thread_get_wp` not quite doing its job. And we of course have two versions - apparently `thread_get_wp'` is the right form and what we need here. We should probably get rid of the first and just keep the `wp'` version, and also for the master.

Reply:
We probably should make an issue so that we don't forget about that.